src/gui/tagfilterwidget.cpp

```diff
@@ -0,0 +1,225 @@
+
+#include "base/bittorrent/session.h"
+#include "base/utils/misc.h"
+#include "autoexpandabledialog.h"
+#include "guiiconprovider.h"
+#include "tagfiltermodel.h"
+#include "tagfilterproxymodel.h"
+
+namespace
+{
+    QString getTagFilter(const TagFilterProxyModel *const model, const QModelIndex &index)
+    {
+        QString tagFilter; // Defaults to All
+        if (index.isValid()) {
+            if (!index.parent().isValid() && (index.row() == 1))
+                tagFilter = "";  // Untagged
+            else if (index.parent().isValid() || (index.row() > 1))
+                tagFilter = model->tag(index);
+        }
+
+        return tagFilter;
+    }
+}
+
+TagFilterWidget::TagFilterWidget(QWidget *parent)
+    : QTreeView(parent)
+{
+    TagFilterProxyModel *proxyModel = new TagFilterProxyModel(this);
+    proxyModel->setSortCaseSensitivity(Qt::CaseInsensitive);
+    proxyModel->setSourceModel(new TagFilterModel(this));
+    setModel(proxyModel);
+    setFrameShape(QFrame::NoFrame);
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+    setUniformRowHeights(true);
+    setHeaderHidden(true);
+    setIconSize(Utils::Misc::smallIconSize());
+#if defined(Q_OS_MAC)
+    setAttribute(Qt::WA_MacShowFocusRect, false);
+#endif
+    setContextMenuPolicy(Qt::CustomContextMenu);
+    sortByColumn(0, Qt::AscendingOrder);
+    setCurrentIndex(model()->index(0, 0));
+
+    connect(this, &TagFilterWidget::collapsed, this, &TagFilterWidget::callUpdateGeometry);
+    connect(this, &TagFilterWidget::expanded, this, &TagFilterWidget::callUpdateGeometry);
+    connect(this, &TagFilterWidget::customContextMenuRequested, this, &TagFilterWidget::showMenu);
+    connect(selectionModel(), &QItemSelectionModel::currentRowChanged, this
+            , &TagFilterWidget::onCurrentRowChanged);
+    connect(model(), &QAbstractItemModel::modelReset, this, &TagFilterWidget::callUpdateGeometry);
+}
+
+QString TagFilterWidget::currentTag() const
+{
+    QModelIndex current;
+    auto selectedRows = selectionModel()->selectedRows();
+    if (!selectedRows.isEmpty())
+        current = selectedRows.first();
+
+    return getTagFilter(static_cast<TagFilterProxyModel *>(model()), current);
+}
+
+void TagFilterWidget::onCurrentRowChanged(const QModelIndex &current, const QModelIndex &previous)
+{
+    Q_UNUSED(previous);
+
+    emit tagChanged(getTagFilter(static_cast<TagFilterProxyModel *>(model()), current));
+}
+
+void TagFilterWidget::showMenu(QPoint)
+{
+    QMenu menu(this);
+
+    QAction *addAct = menu.addAction(
+        GuiIconProvider::instance()->getIcon("list-add")
+        , tr("Add tag..."));
+    connect(addAct, &QAction::triggered, this, &TagFilterWidget::addTag);
+
+    auto selectedRows = selectionModel()->selectedRows();
+    if (!selectedRows.empty() && !TagFilterModel::isSpecialItem(selectedRows.first())) {
```

Comment:
I was going for consistency with Categories, which also omits the "Remove Category" option from All / Uncategorized. So if one is changed, we should probably change both.